cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/metadata/EntityMetadataEmitterContext.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.etl.api.metadata;
+
+import co.cask.cdap.api.metadata.MetadataEntity;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Interfaces for methods exposed in pipeline to write metadata. This writer is tied to an
+ * entity and can only write metadata for the specific entity or its derivatives
```

Comment:
If this is tied to an entity, how can I create one of  these contexts?

Also  I am not sure why it is called a context. Isn't it just a MetadataEmitter? 

Also, it can remove metadata, too. Is Emitter a good name then?

Reply:
The user will not create it. The source and sink will have an implementation which is created by using the dataset name which they are tied too. 
I am open to suggestions on name.

Reply:
How can I write a custom plugin that records meta data for the input or output of the pipeline? Assume a user who relies on standard sources and sinks, but, on success of the pipeline runs wants to record metrics for the source and propagate metadata from the source to the sink. The source and sink would not record that meta data, so he needs a custom action plugin.